Design-concepts.md

```diff
@@ -260,6 +264,554 @@ For FuseML to be functional, the following design elements and components are ne
+
+The following high-level composable pipeline segment templates are featured in this exercise:
+* builder segment: takes in an MLFlow compatible codeset artifact as input and generates a runnable artifact representing the environment that can be used to run any of its entry points. This is based on the `mlflow-builder` runnable installed by the MLFlow extension
+* trainer segment: takes in an MLFlow compatible codeset artifact as input, runs it and outputs the resulted trained ML model. This segment is based on the runnable generated at runtime by the builder segment
+* predictor segment: takes in a trained ML model as input, starts a prediction service for it and outputs its exposed endpoint (i.e. the URL where the service receives HTTP calls for prediction). This is based on the `kfserving-predictor` runnable installed by the KFServing extension
+
+Each of these segments can be instantiated and executed independently, as long as there are available inputs (e.g. compatible codesets and models) that they can consume and as long as the values of all other input parameters are explicitly specified (or have default values). The steps detailed in the next part will cover defining the individual pipeline templates and then building a master end-to-end pipeline composed of all three segments which is automatically run according to the available MLFlow compatible codeset versions published as detailed in the _Publishing Codesets_ section. 
+
+##### Inline Pipeline Templates
+
+Before going into how runnables can be used as building blocks for pipelines, here's what the complete end-to-end pipeline may look like _without_ the use of runnables as composable components:
+
+  ```
+  pipeline-template:
+    name: mlflow-kfserving-e2e
+    description: |
+      End-to-end pipeline template that takes in an MLFlow compatible codeset,
+      runs the MLFlow project to train a model, then creates a KFServing prediction 
+      service that can be used to run predictions against the model.
+    inputs:
+      - name: mlflow-codeset
+        description: an MLFlow compatible codeset
+        type: codeset
+        labels:
+          - mlflow-project
+      - name: predictor
+        description: type of predictor engine
+        type: string
+        default: auto
+    outputs:
+      - name: prediction-url
+        description: the URL where the exposed prediction service endpoint can
+        be contacted to run predictions 
+        type: string
+    steps:
+      - name: builder
+        image:
+          registryURL: ghcr.io/fuseml
+          repository: mlflow-builder
+          tag: 1.0
+        inputs:
+          - name: mlflow-codeset
+            codeset:
+              path: /project
+        outputs:
+          - name: mlflow-env
+            image:
+              fromfile: /project/.fuseml/Dockerfile
+              repository: "mlflow-builder/{{inputs[mlflow-codeset].name}}"
+              tag: "{{inputs[mlflow-codeset].version}}"
```

Comment:
Is there a reason for having a `name` for a codeset `input`? I think that specifying a codeset input would be clearer without the name, for example:
```inputs:
  - codeset:
      name: {{inputs.mlflow-codeset}}
      path: /project
```
It might even make sense to remove the `path` parameter and hardcode a path were we mount the codeset (unless there are cases where the codeset needs to be mounted in a custom path)

Reply:
I agree on clarity of using `inputs[<codesetname>]`. This however means that codeset inputs need to be dict instead of a list, which is fine, IMO.

Reply:
No, there's no reason for having names for step inputs. This only becomes valuable when the steps themselves can be detached as composable and reusable components, but those are runnables and are not in this scope.
Regarding the `path` parameter, sure, we can use a convention and define a hardcoded path where inputs are mounted, but this creates a situation that makes it more difficult to reuse existing container images with fuseml without modifications. IMHO however, it should be the containers that indicate and control the paths where they expect inputs and where they generate outputs, not the pipeline framework, to ensure portability between frameworks. At a minimum, this should be modeled as another parameter.